src/query/ee/src/storages/fuse/operations/ngram_index.rs

```diff
@@ -0,0 +1,315 @@
+use databend_common_storages_fuse::operations::MutationLogs;
+use databend_common_storages_fuse::operations::TableMutationAggregator;
+use databend_common_storages_fuse::FuseStorageFormat;
+use databend_common_storages_fuse::FuseTable;
+use databend_query::storages::index::NgramArgs;
+use databend_storages_common_cache::LoadParams;
+use databend_storages_common_io::ReadSettings;
+use databend_storages_common_table_meta::meta::BlockMeta;
+use databend_storages_common_table_meta::meta::ExtendedBlockMeta;
+use databend_storages_common_table_meta::meta::Location;
+use databend_storages_common_table_meta::meta::Statistics;
+use databend_storages_common_table_meta::meta::Versioned;
+use opendal::Operator;
+
+pub async fn do_refresh_ngram_index(
+    fuse_table: &FuseTable,
+    ctx: Arc<dyn TableContext>,
+    _index_name: String,
+    segment_locs: Option<Vec<Location>>,
+    pipeline: &mut Pipeline,
+) -> Result<()> {
+    let Some(snapshot) = fuse_table.read_table_snapshot().await? else {
+        // no snapshot
+        return Ok(());
+    };
+
+    let table_schema = &fuse_table.get_table_info().meta.schema;
+    // Read blocks in segment.
+    let projection = Projection::Columns((0..table_schema.fields().len()).collect());
+
+    let block_reader =
+        fuse_table.create_block_reader(ctx.clone(), projection, false, false, false)?;
+
+    let segment_reader =
+        MetaReaders::segment_info_reader(fuse_table.get_operator(), table_schema.clone());
+
+    // If no segment locations are specified, iterates through all segments
+    let segment_locs = if let Some(segment_locs) = segment_locs {
+        segment_locs
+            .into_iter()
+            .filter(|s| snapshot.segments.contains(s))
+            .collect()
+    } else {
+        snapshot.segments.clone()
+    };
+
+    if segment_locs.is_empty() {
+        return Ok(());
+    }
+    let operator = fuse_table.get_operator_ref();
+
+    // Read the segment infos and collect the block metas that need to generate the index.
+    let mut block_metas = VecDeque::new();
+    let mut block_meta_index_map = HashMap::new();
+    for (segment_idx, (segment_loc, ver)) in segment_locs.into_iter().enumerate() {
+        let segment_info = segment_reader
+            .read(&LoadParams {
+                location: segment_loc.to_string(),
+                len_hint: None,
+                ver,
+                put_cache: false,
+            })
+            .await?;
+
+        for (block_idx, block_meta) in segment_info.block_metas()?.into_iter().enumerate() {
+            let index_location =
+                TableMetaLocationGenerator::gen_bloom_index_location_from_block_location(
+                    &block_meta.location.0,
+                );
+            // only generate bloom index if it is not exist.
+            if (operator.stat(&index_location).await).is_err()
+                || block_meta.ngram_filter_index_size.is_none()
```

Comment:
Only checking the index size might be incorrect. User might refresh the index after modifying the options.